packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/config.ts

```diff
@@ -0,0 +1,363 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import * as path from 'path';
+import { Runtime } from './runtime';
+
+/**
+ * Custom resource provider types.
+ */
+export enum ProviderType {
+  /**
+   * `CdkFunction`
+   */
+  CDK_FUNCTION = 'CdkFunction',
+
+  /**
+   * `CdkSingletonFunction`
+   */
+  CDK_SINGLETON_FUNCTION = 'CdkSingletonFunction',
+
+  /**
+   * `CdkCustomResourceProvider`
+   */
+  CDK_CUSTOM_RESOURCE_PROVIDER = 'CdkCustomResourceProvider',
+
+  /**
+   * Do not create a custom resource provider.
+   *
+   * Note: This is used to just move source code for airlifting.
+   */
+  CDK_NO_OP = 'CdkNoOp',
+}
+
+/**
+ * Properites used to generate individual custom resource providers.
+ */
+export interface ProviderProps {
+  /**
+   * The custom resource provider type to generate.
```

Comment:
You need to explain the rules of thumb that go into this, otherwise the person after you is going to be mighty confused.

How would you choose between the 4 different alternatives?